components/places/src/storage/history.rs

```diff
@@ -1019,8 +1019,20 @@ pub mod history_sync {
     }
 
     pub fn apply_synced_deletion(db: &PlacesDb, guid: &SyncGuid) -> Result<()> {
+        // First we delete any visits for the page
+        // because it's possible the moz_places foreign_count is not 0
+        // and thus the moz_places entry won't be deleted.
         db.execute_cached(
-            "DELETE FROM moz_places WHERE guid = :guid",
+            "DELETE FROM moz_historyvisits
+              WHERE place_id IN (
+                  SELECT id
+                  FROM moz_places
+                  WHERE guid = :guid
+              )",
+            &[(":guid", guid)],
+        )?;
+        db.execute_cached(
```

Comment:
I verified that if we had bookmarks referencing the place_id and and applied a tombstone, we do get a `FOREIGN KEY constraint failed` error back from sqlite

Reply:
Nice, thanks for checking!